src/nvim/buffer.c

```diff
@@ -3904,31 +3904,53 @@ int build_stl_str_hl(
         if (item[item_idx].type == Middle) {
           middle_mark = item_idx;
           break;
-          /*
-          // Move the statusline to make room for the middle characters
-          char_u *middle_end = item[item_idx].start + (maxwidth - width);
-          STRMOVE(middle_end, item[item_idx].start);
+        }
+      }
+
+      if (middle_mark > 0) {
+        // Handle a combination of middle mark and centering
+        //
+        // text\ %[this\ text\ here%=\ and\ some%]finishing
+        // text      his text here and some     finishing
+        // quad1             quad2|quad3            quad4
+        //       --fill1--                --fill2--
+        // int len_a = item[center_begin].start - out;
+        int len_b = item[middle_mark].start - out;
+        // int len_c = item[center_end].start - out;
+        // Not sure about this one
+        int len_d = width;
+
+        int fill_left = maxwidth / 2 - len_b;
+        int fill_right = maxwidth - maxwidth / 2 - len_d + len_b;
```

Comment:
Personally I'd always be explicit in such situations or at least use an visual style to make the operator precedence clearer, so `a - (b / c) - d + e` or `a - b/c - d +e`. This case is simple enough and doesn't need to be changed, but for the future it might be nice. ;-)